shared/js/dataDriver.js

```diff
@@ -0,0 +1,283 @@
+      return 2147483647;
+    }
+
+    /**
+     * Get the server time for a local time or now.
+     */
+    getServerTime(time = Date.now()) {
+      return (this._clockSkew || 0) + time;
+    }
+
+    /**
+     * Generate an id based on the time.
+     */
+    makeId(time = this.getServerTime()) {
+      // Call the function that has locally scoped state.
+      return makeId(time);
+    }
+
+    /**
+     * Request the most recent chat messages between two times.
+     */
+    requestChat(startTime = 0, endTime = this.getServerTime(), limit = 0) {
+      let query = this._buildQuery({
+        orderBy: '"$key"',
+        // Fudge the time to get an inclusive range with the random parts.
+        startAt: `"chat!${this.makeId(Math.max(0, startTime - 1))}"`,
+        endAt: `"chat!${this.makeId(endTime + 1)}"`,
+        // Always have a limit to get things sorted correctly.
+        limitToLast: limit > 0 ? limit : this.MAX_LIMIT
+      });
+      return this._request("GET", this._buildUrl(), query);
+    }
+
+    /**
+     * Update a record additionally storing the timestamp.
+     */
+    update(type, id, value) {
+      let key = `${type}!${id}`;
+      return this._request("PUT", this._buildUrl(key), {
+        "timestamp": {
+          ".sv": "timestamp"
+        },
+        value: value
+      });
+    }
+
+    /** ****************** **
+     *** Internal Helpers ***
+     ** ****************** **/
+
+    /**
+     * Convert an object with key/value pairs into a query string.
+     */
+    _buildQuery(object) {
+      return Object.keys(object).map(key => `${key}=${object[key]}`).join("&");
+    }
+
+    /**
+     * Get a REST endpoint for a resource defaulting to the current room.
+     */
+    _buildUrl(resource = "", roomToken = this._roomToken) {
+      return `${this.BASE_FIREBASE}/${roomToken}/${resource}.json`;
+    }
+
+    /**
+     * Connect to a specified room for streaming updates and future requests.
+     */
+    _connectToRoom(token) {
+      this._roomToken = token;
+      this._sse = new EventSource(this._buildUrl() + "?" + this._buildQuery({
+        limitToLast: 1,
+        orderBy: '"timestamp"'
+      }));
+      this._sse.addEventListener("open", this);
+      this._sse.addEventListener("put", this);
+
+      // Determine the clock skew before continuing initialization.
+      this.update("meta", "lastConnect").then(({ timestamp }) => {
+        this._clockSkew = timestamp - Date.now();
+      }).then(() => {
+        let now = this.getServerTime();
+        this.requestChat(now - 24 * 60 * 60 * 1000, now, 25).then(result =>
+          this._processRecords(result));
+      });
+    }
+
+    /**
+     * Handle a single record dispatching appropriate actions.
+     */
+    _processRecord(key, data) {
+      let [, type] = key.match(/^([^!]+)!(.+)$/);
+      switch (type) {
+        case "chat": {
+          let chat = data.value;
+          chat.receivedTimestamp = new Date(data.timestamp).toISOString();
```

Comment:
We should probably validate data.timestamp as well - or at least check that if something is wrong here, we a) log it b) don't break other functionality.